src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs

```diff
@@ -11743,5 +11743,1308 @@ public void ReturnThis_02(LanguageVersion languageVersion)
+        }
+
+        [Fact]
+        public void RefInitializer_OnArray()
+        {
+            var source = @"
+public ref struct C
+{
+    C M()
+    {
+        int x = 0;
+        var c = new C { array = { [0] = ref x } }; // 1
+        return c;
+    }
+
+    void M2()
+    {
+        int x = 0;
+        C c2 = new C();
+        c2.array[0] = ref x; // 2
+    }
+
+    public int[] array;
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (7,35): error CS8373: The left-hand side of a ref assignment must be a ref variable.
+                //         var c = new C { array = { [0] = ref x } }; // 1
+                Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "[0]").WithLocation(7, 35),
+                // (15,9): error CS8373: The left-hand side of a ref assignment must be a ref variable.
+                //         c2.array[0] = ref x; // 2
+                Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "c2.array[0]").WithLocation(15, 9)
+                );
+        }
+
+        [Fact]
+        public void RefInitializer_OnPointer()
+        {
+            var source = @"
+public unsafe class C
+{
+    public int* pointer;
+
+    C M()
+    {
+        int x = 0;
+        var c = new C { pointer = { [0] = ref x } }; // 1
+        return c;
+    }
+
+    void M2()
+    {
+        int x = 0;
+        C c2 = new C();
+        c2.pointer[0] = ref x; // 2
+    }
+}
+";
+            var comp = CreateCompilation(source, options: TestOptions.UnsafeDebugDll);
+            comp.VerifyDiagnostics(
+                // (9,37): error CS8373: The left-hand side of a ref assignment must be a ref variable.
+                //         var c = new C { pointer = { [0] = ref x } }; // 1
+                Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "[0]").WithLocation(9, 37),
+                // (17,9): error CS8373: The left-hand side of a ref assignment must be a ref variable.
+                //         c2.pointer[0] = ref x; // 2
+                Diagnostic(ErrorCode.ERR_RefLocalOrParamExpected, "c2.pointer[0]").WithLocation(17, 9)
+                );
+        }
+
+        [Fact]
+        public void RefInitializer_OnEvent()
+        {
+            var source = @"
+int x = 42;
+var r = new C { a = ref x }; // 1
+
+C c = default;
+c.a = ref x; // 2
+
+class C
+{
+    public event System.Action a;
+}
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (3,17): error CS0070: The event 'C.a' can only appear on the left hand side of += or -= (except when used from within the type 'C')
+                // var r = new C { a = ref x }; // 1
+                Diagnostic(ErrorCode.ERR_BadEventUsage, "a").WithArguments("C.a", "C").WithLocation(3, 17),
+                // (6,3): error CS0070: The event 'C.a' can only appear on the left hand side of += or -= (except when used from within the type 'C')
+                // c.a = ref x; // 2
```

Comment:
might be nice to verify the behavior here on e.g. `void M() { this.a = ref x; }` or `void M() { this.a += ref x; }`.